mlsdbtool/main.c

```diff
@@ -0,0 +1,180 @@
+tail -n +2 [CSV file] | sort -t, -k2n,2 -k3n,3 -k4n,4 -k5n,5 -k1,1 | geoclue-mlsdb-tool
+or (preferrably) use the wrapper script.
+
+---
+
+The "network" is allocated as follows:
+0000000000000000000000000000000000000000000000000000000000000000
+        \____ ___/\_______ ______/\_____________ ____________/\/
+             v            v                     v              v
+          NET: 10b Area: 16 bits        Cell ID: 28 bits     Radio: 2b 
+
+The "position" is allocated as 2 concatenated 32-bit floats
+with longitude first and then latitude.
+
+This program will produce 2 files per mcc: .ntw for network data and .loc
+for location data. To finalize the mlsdb data file the two should simply
+be concatenated into a .dat file. (The wrapper will do this too.)
+
+Since the network portion and the data portion are exactly the same size,
+the file is simply split so the the 1st half contains network data, and the
+corresponding location data is found in ${net_data_pos} + ${file_size} / 2
+*/
+uint64_t network;
+uint64_t position;
+uint64_t previous = 0;
+
+// Helper function for debugging (prints out a 64b int as binary)
+void print_bin(uint64_t n) {
+    if (n > 1) {
+        print_bin(n >> 1);
+    }
+    printf("%lu", n & 1);
+}
+
+void add_network(char *str, size_t shift) {
+    uint64_t t = atoi(str);
+    t <<= shift;
+    network |= t;
+}
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstrict-aliasing"
+void add_position(char *str, size_t shift) {
+    float f = atof(str);
+    // Pretend the 32b float is a 32b int, and assign the value to a 64b int.
+    uint64_t t = *(uint32_t*)&f;
+    t <<= shift;
+    position |= t;
+}
+#pragma GCC diagnostic pop
+
+void add_radio(char *str) {
+    switch(str[0]) {
+    case 'G': // GSM
+        break;
+    case 'L': // LTE
+        ++network;
+        break;
+    case 'U': // UMTS
+        network += 2;
+        break;
+    default:
+        network += 3;
+        break;
+    }
+}
+
+int main() {
+    char c;
+    char line[150];
+    char mcc_fn[16];
+    char loc_fn[16];
+    size_t pos = 0, count = 0, mcc_old = 0, mcc_num = 0, mcc_p = 0, net_p = 0, area_p = 0, cell_p = 0, lon_p = 0, lat_p = 0;
+    FILE *fp_mcc = NULL, *fp_loc = NULL;
+    while (read(STDIN, &c, 1) > 0) {
+        if (c == NEWLINE) {
+            pos = 0;
+            count = 0;
+            position = 0;
+            network = 0;
+            mcc_num = atoi(&line[mcc_p]);
+            // We have a new mcc; we need to open new files.
+            if (mcc_num != mcc_old) {
+                if (mcc_old > 0) {
+                    fclose(fp_mcc);
+                    fclose(fp_loc);
+                }
+                strcpy(mcc_fn, "./");
+                strcpy(loc_fn, "./");
+                strcat(mcc_fn, &line[mcc_p]);
+                strcat(loc_fn, &line[mcc_p]);
+                strcat(mcc_fn, ".ntw");
+                strcat(loc_fn, ".loc");
```

Comment:
Two `snprintf()`s might be overall easier to read here as well as provide security of avoiding any potential buffer overflow.